### ISSUE TITLE
WC_Data::apply_changes usage of array_merge()

### ISSUE THREAD

**Issue by james-allan** [woocommerce/woocommerce#13509]
## EXPLANATION OF THE ISSUE

Calling `$order->save()` on an order with no billing details causes future attempts to set billing props to fail (do nothing). 

## STEPS TO REPRODUCE THE ISSUE
Example code snippet: 

```php
$order = wc_create_order( array( 'customer_id' => 1 ) );

$order->set_billing_last_name( 'Smith' );
$order->save();
``` 

------------------
`wc_create_order()` calls `$order->save()` which eventually leads to `WC_Data::apply_changes()` which calls `array_merge()`. 

The result of the [`array_merge`](https://github.com/woocommerce/woocommerce/blob/master/includes/abstracts/abstract-wc-data.php#L553) inside `WC_Data::apply_changes()` is: 

```
Array
(
    [parent_id] => 0
    [status] => pending
    [currency] => AUD
    [version] => 2.7.0
    [prices_include_tax] => 1
    [date_created] => 1489013775
    [date_modified] => 
    [discount_total] => 0
    [discount_tax] => 0
    [shipping_total] => 0
    [shipping_tax] => 0
    [cart_tax] => 0
    [total] => 0
    [total_tax] => 0
    [customer_id] => 1
    [order_key] => wc_order_58bfff6fa0b
    [billing] => Array
        (
            [email] => james@prospress.com
        )

    [shipping] => Array
        (
            [first_name] => 
            [last_name] => 
            [company] => 
            [address_1] => 
            [address_2] => 
            [city] => 
            [state] => 
            [postcode] => 
            [country] => 
        )

    [payment_method] => 
    [payment_method_title] => 
    [transaction_id] => 
    [customer_ip_address] => ::1
    [created_via] => 
    [customer_note] => 
    [date_completed] => 
    [date_paid] => 
    [cart_hash] => 
)
```
_Notice the `billing` array only contains the email key/value. (This comes from the `$this->maybe_set_user_billing_email();` inside `WC_Order->save()`). If you're interested in the the original `$data` and `$changes` contents, you can find them [here](https://cloudup.com/czm1SxVglWA)_

Because the `$order->data` variable no longer contains all the billing address keys, calling any `$order->set_address_prop()` with any billing key won't work. (Because of this condition [class-wc-order.php#L830](https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-order.php#L830))

Basically, I think this [`array_merge`](https://github.com/woocommerce/woocommerce/blob/master/includes/abstracts/abstract-wc-data.php#L553) should be replaced with an `array_replace_recursive` or similar so that multidimensional arrays maintain all keys.

**Comment by mikejolley** [comment]
@claudiulodro If you want this one, it could use a unit test too :)
